### ISSUE TITLE
[Feature Request] Support monorepos

### ISSUE THREAD

**Issue by mjsisley** [frontsideair/yarnhook#14]
Thanks for the great lib!

Currently, doesn't detect changes in nested `package.json` and `yarn.lock`. What do you think about adding support for monorepos?

**Comment by frontsideair** [comment]
I'd love it, but I don't have any experience with monorepos. I'd love a PR, but if you can outline a plan that would be good too, and I might get to do it when I have time.

**Comment by mjsisley** [comment]
This tool supports monorepos https://github.com/renovateapp/renovate and could be a reference.

Basically, a monorepo has subdirectories that also contain package.json and lockfiles. Traversing the sub-directories (ignoring node_modules) and performing the code already performed in this repo in any subdirectory that has a package.json and a lockfile should do the trick.

**Comment by frontsideair** [comment]
It turned out to be trickier than I imagined. `lerna bootstrap` installs dependencies of subpackages but it doesn't install its own dependencies, so running `yarn install` is necessary.

Also the current strategy for detecting a diff works only for a lockfile adjacent to a `package.json` so deeper changes are not detected. I'll think more about this.

The current code is in `lerna` branch, feel free to check it out.

**Comment by frontsideair** [comment]
I found how `renovate` handles monorepos: https://github.com/renovateapp/renovate/blob/fc4d5fc08d966c54e319c42e9acb9da89c0dd0df/lib/manager/npm/monorepos.js

Looks like it currently supports `lerna` and `yarn workspaces`.

**Comment by cinderblock** [comment]
What if there was an optional argument or `yarnhook.json` config file that specified a list of directories or globs to scan for changes? Possibly recursively loading `yarnhook.json` files in found directories?

**Comment by Skylion007** [comment]
Yeah, just being able to detect an package.json in a subfolder would be great!

**Comment by Stanzilla** [comment]
any news here?

**Comment by frontsideair** [comment]
@Stanzilla thanks for asking!

I've neglected this project for too long, but had many ideas brewing and didn't have the throughput to realize them. These ideas are fine as their own projects, but this project still requires some attention.

Regarding this specific issue, contributions are welcome and I'd like to review and merge them. Since I don't have any monorepos that I maintain (apart from this one) I'm not the best person to implement this.

The barrier to entry may seem high, but the meat of the project is very small and I can try to help along the way.
